vdaf/prio3/internal/flp/flp.go

```diff
@@ -0,0 +1,144 @@
+// Package flp provides a Fully-Linear Proof (FLP) system.
+package flp
+
+import (
+	"errors"
+
+	"github.com/cloudflare/circl/vdaf/prio3/arith"
+	"github.com/cloudflare/circl/vdaf/prio3/internal/cursor"
+)
+
+// FLP is an instance of a FLP by Boneh et al. Crypto, 2019 paper
+// "Zero-Knowledge Proofs on Secret-Shared Data via Fully Linear PCPs",
+// https://ia.cr/2019/188
+// plus some changes described in VDAFs CFRG draft.
+// https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-vdaf-13#section-7.1
+type FLP[
+	G Gadget[P, V, E, F],
+	P arith.Poly[P, E], V arith.Vec[V, E], E arith.Elt, F arith.Fp[E],
+] struct {
+	// Eval evaluates the arithmetic circuit on a measurement and joint randomness.
+	Eval func(out V, g Gadget[P, V, E, F], numCalls uint, meas, jointRand V, shares uint8)
```

Comment:
both the gadget (either a `ProveGadget` or `QueryGadget`) and the `numGadgetCalls` are passed as a side effect that Go doesn't support abstract classes. Think of this function signature to be an externally-provided callback.

this may need refactor to add more generality on future VDAFs.